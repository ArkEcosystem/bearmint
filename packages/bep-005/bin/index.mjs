#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
