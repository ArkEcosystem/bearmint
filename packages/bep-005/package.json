{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
