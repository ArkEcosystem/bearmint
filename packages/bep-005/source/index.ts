export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
