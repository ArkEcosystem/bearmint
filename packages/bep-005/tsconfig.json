{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
