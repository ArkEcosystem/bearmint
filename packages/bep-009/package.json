{
	"name": "@bearmint/bep-009",
	"version": "0.0.0",
	"description": "This library is the official reference implementation for BEP-009 (https://bep.dev/009)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"files": [
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch",
		"docs": "typedoc source",
		"knip": "knip",
		"lint": "eslint 'source/**/*.ts' --format=codeframe --config=../../config/eslintrc.js --fix",
		"test": "vitest run",
		"test:coverage": "vitest run --coverage",
		"test:ui": "vitest --ui",
		"test:watch": "vitest"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@ethereumjs/rlp": "^4.0.0",
		"@noble/hashes": "1.1.3",
		"camelcase": "^7.0.0",
		"camelcase-keys": "^8.0.2",
		"dayjs": "1.11.6",
		"dot-prop": "^7.2.0",
		"env-paths": "^3.0.0",
		"matcher": "^5.0.0",
		"moize": "6.1.3",
		"semver": "7.3.8",
		"snake-case": "^3.0.4",
		"super-regex": "^0.2.0"
	},
	"devDependencies": {
		"@bearmint/bep-005": "workspace:^",
		"@types/node": "18.11.9",
		"@types/semver": "7.3.13"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "public"
	}
}
