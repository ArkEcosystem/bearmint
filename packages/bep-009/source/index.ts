export * from './assert.js'
export * from './base64.js'
export * from './big-number.js'
export * from './bytes.js'
export * from './camel-case.js'
export * from './directory.js'
export * from './dot-property.js'
export * from './environment-paths.js'
export * from './file.js'
export * from './glob.js'
export * from './hash.js'
export * from './hex.js'
export * from './is.js'
export * from './json.js'
export * from './non-nullable-map.js'
export * from './number.js'
export * from './numbers.js'
export * from './package.js'
export * as Paths from './paths.js'
export * from './regex.js'
export * as RLP from './rlp.js'
export * from './semver.js'
export * from './snake-case.js'
export * from './string.js'
export * from './uniq.js'
export * from './zeit.js'
