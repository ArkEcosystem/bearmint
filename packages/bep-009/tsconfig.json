{
	"compilerOptions": {
		"outDir": "distribution",
		"typeRoots": ["node_modules/@types"]
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
