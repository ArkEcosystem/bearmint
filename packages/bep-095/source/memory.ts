import { hexToBytes, zeit } from '@bearmint/bep-009'
import { PROTO_OPTS } from '@bearmint/bep-013'
import { SignedTx } from '@bearmint/bep-018'

import type { Signature, StoreTx } from './database.js'
import type { MuSig } from './types.js'

type TxIndex = Record<string, string[]>

interface SignatureRequest {
	hash: string
	muSig: MuSig
	signature: Signature
}

export function makeMemory({ pendingLimit }: { pendingLimit: number }) {
	const txs: Record<string, StoreTx> = {}
	const txIdsBySender: TxIndex = {}
	const txIdsByPublicKey: TxIndex = {}

	function restoreTx(tx: StoreTx): string {
		const { sender } = SignedTx.fromBinary(hexToBytes(tx.hash), PROTO_OPTS)

		const pendingCount: number = (txIdsBySender[sender] ?? []).length

		if (pendingCount >= pendingLimit) {
			throw new Error(
				`The public key [${sender}] has reached its max of ${pendingLimit} pending txs.`,
			)
		}

		txs[tx.hash] = tx

		if (txIdsBySender[sender] === undefined) {
			txIdsBySender[sender] = []
		}

		txIdsBySender[sender]?.push(tx.hash)

		for (const publicKey of tx.muSig.slaves) {
			txIdsByPublicKey[publicKey] = txIdsByPublicKey[publicKey] ?? []
			txIdsByPublicKey[publicKey]?.push(tx.hash)
		}

		return tx.hash
	}

	function getTxById(storeId: string) {
		return txs[storeId]!
	}

	function removeById(storeId: string): void {
		const { hash, muSig } = txs[storeId]!
		const { sender } = SignedTx.fromBinary(hexToBytes(hash), PROTO_OPTS)

		// removes indexes
		txIdsBySender[sender] = txIdsBySender[sender]?.filter((id) => id !== storeId) ?? []

		for (const publicKey of muSig.slaves) {
			txIdsByPublicKey[publicKey] =
				txIdsByPublicKey[publicKey]?.filter((id) => id !== storeId) ?? []
		}

		// remove actual tx
		delete txs[storeId]
	}

	return {
		deleteAllTxs() {
			for (const id of Object.keys(txs)) {
				removeById(id)
			}
		},
		getAllTxs() {
			return Object.values(txs)
		},
		getTxById,
		getTxStoreByPublicKey(publicKey: string) {
			const storeIdsBySender = txIdsBySender[publicKey] ?? []
			const storeIdsByPublicKey = txIdsByPublicKey[publicKey] ?? []

			const allById: Record<string, StoreTx> = {}
			for (const id of [...storeIdsBySender, ...storeIdsByPublicKey]) {
				allById[id] = getTxById(id)
			}

			return Object.values(allById)
		},
		loadTxs(storeTxs: StoreTx[]) {
			for (const tx of storeTxs) {
				restoreTx(tx)
			}
		},
		removeById,
		saveTx({ hash, muSig, signature }: SignatureRequest): string {
			return restoreTx({
				hash,
				muSig,
				signatures: [signature],
				timestamp: zeit.toUnix(),
			})
		},
		updateTx({ hash, signature }: SignatureRequest) {
			txs[hash]?.signatures.push(signature)
		},
	}
}

export type Memory = ReturnType<typeof makeMemory>
