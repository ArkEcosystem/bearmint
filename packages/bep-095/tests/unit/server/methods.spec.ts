import { expect, describe, it, beforeEach, afterEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import { bytesToHex, hexToBytes } from '@bearmint/bep-009'
import { SignedTx } from '@bearmint/bep-018'
import BLS from 'bls-eth-wasm'
import MockDate from 'mockdate'

import type { Memory } from '../../../source/memory.js'
import { makeMemory } from '../../../source/memory.js'
import {
	makeDeleteHandler,
	makeListHandler,
	makeShowHandler,
	makeUpdateHandler,
} from '../../../source/server/methods.js'
import { accounts, data, muSig, muSigFull } from '../../fixture.js'
import type { MuSigFactory } from '../../musig.js'
import { makeMuSigFactory } from '../../musig.js'

async function assertPending({ context, hash }) {
	await expect(async () => context.subject.show.execute({ hash })).rejects.toThrow(
		'The tx is not ready yet.',
	)
}

async function signHash({ account, hash, muSigFactory }) {
	return {
		id: account.id,
		publicKey: account.publicKey,
		signature: bytesToHex(
			await muSigFactory.sign({
				hash: hexToBytes(hash),
				signatory: {
					id: Buffer.from(account.id, 'hex'),
					privateKey: Buffer.from(account.privateKey, 'hex'),
				},
			}),
		),
	}
}

describe<{
	hash: string
	subject: {
		delete: { execute: CallableFunction; schema: object }
		list: { execute: CallableFunction; schema: object }
		show: { execute: CallableFunction; schema: object }
		store: { execute: CallableFunction; schema: object }
	}
	muSigFactory: MuSigFactory
	memory: Memory
}>('Server', () => {
	beforeEach(async (context) => {
		await BLS.init(BLS.BLS12_381)

		context.hash = bytesToHex(new SignedTx(data).toBinary())
		context.memory = makeMemory({ pendingLimit: 3 })
		context.muSigFactory = makeMuSigFactory()

		context.subject = {
			delete: makeDeleteHandler(context.memory),
			list: makeListHandler(context.memory),
			show: makeShowHandler(context.memory),
			store: makeUpdateHandler(context.memory),
		}

		MockDate.set(Date.parse('2022-05-09T13:14:01+00:00'))
	})

	afterEach(() => {
		MockDate.reset()
	})

	it('should fail if there are less than M signatures', async (context) => {
		await context.subject.store.execute({
			hash: context.hash,
			muSig,
			signature: await signHash({
				account: accounts[0],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		await assertPending({ context, hash: context.hash })
	})

	it('should complete an M-of-N scheme', async (context) => {
		expect(
			(
				await context.subject.store.execute({
					hash: context.hash,
					muSig,
					signature: await signHash({
						account: accounts[0],
						hash: context.hash,
						muSigFactory: context.muSigFactory,
					}),
				})
			).hash,
		).toStrictEqual(context.hash)

		await assertPending({ context, hash: context.hash })

		expect(
			(
				await context.subject.store.execute({
					hash: context.hash,
					muSig,
					signature: await signHash({
						account: accounts[1],
						hash: context.hash,
						muSigFactory: context.muSigFactory,
					}),
				})
			).hash,
		).toStrictEqual(context.hash)

		const responseShow = await context.subject.show.execute({
			hash: context.hash,
		})

		expect(responseShow).toMatchSnapshot()
	})

	it('should complete an N-of-N scheme', async (context) => {
		const responseStore = await context.subject.store.execute({
			hash: context.hash,
			muSig: muSigFull,
			signature: await signHash({
				account: accounts[0],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		expect(responseStore.hash).toStrictEqual(context.hash)

		await assertPending({ context, hash: context.hash })

		expect(
			(
				await context.subject.store.execute({
					hash: context.hash,
					muSig: muSigFull,
					signature: await signHash({
						account: accounts[1],
						hash: context.hash,
						muSigFactory: context.muSigFactory,
					}),
				})
			).hash,
		).toStrictEqual(context.hash)

		await assertPending({ context, hash: context.hash })

		expect(
			(
				await context.subject.store.execute({
					hash: context.hash,
					muSig: muSigFull,
					signature: await signHash({
						account: accounts[2],
						hash: context.hash,
						muSigFactory: context.muSigFactory,
					}),
				})
			).hash,
		).toStrictEqual(context.hash)

		expect(
			await context.subject.show.execute({
				hash: context.hash,
			}),
		).toMatchSnapshot()
	})

	it('should respond with code -32004 if the tx does not exist for [show]', async (context) => {
		await expect(async () =>
			context.subject.show.execute({
				hash: context.hash,
			}),
		).rejects.toThrow('-32004')
	})

	it('should delete a tx by its hash', async (context) => {
		expect(
			(
				await context.subject.store.execute({
					hash: context.hash,
					muSig: muSigFull,
					signature: await signHash({
						account: accounts[0],
						hash: context.hash,
						muSigFactory: context.muSigFactory,
					}),
				})
			).hash,
		).toStrictEqual(context.hash)

		expect(
			(
				await context.subject.delete.execute({
					hash: context.hash,
				})
			).hash,
		).toStrictEqual(context.hash)

		await expect(() =>
			context.subject.show.execute({
				hash: context.hash,
			}),
		).rejects.toThrow('-32004')
	})

	it('should respond with code -32004 if the tx does not exist for [delete]', async (context) => {
		await expect(() =>
			context.subject.delete.execute({
				hash: context.hash,
			}),
		).rejects.toThrow('-32004')
	})

	it('should list all txs', async (context) => {
		await context.subject.store.execute({
			hash: context.hash,
			muSig,
			signature: await signHash({
				account: accounts[0],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		expect(
			await context.subject.list.execute({
				publicKey: muSig.master,
			}),
		).toMatchSnapshot()
	})

	it('should list pending txs', async (context) => {
		await context.subject.store.execute({
			hash: context.hash,
			muSig,
			signature: await signHash({
				account: accounts[0],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		expect(
			await context.subject.list.execute({
				publicKey: muSig.master,
				state: 'pending',
			}),
		).toMatchSnapshot()
	})

	it('should list ready txs', async (context) => {
		await context.subject.store.execute({
			hash: context.hash,
			muSig,
			signature: await signHash({
				account: accounts[0],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		await context.subject.store.execute({
			hash: context.hash,
			muSig,
			signature: await signHash({
				account: accounts[1],
				hash: context.hash,
				muSigFactory: context.muSigFactory,
			}),
		})

		expect(
			await context.subject.list.execute({
				publicKey: muSig.master,
				state: 'ready',
			}),
		).toMatchSnapshot()
	})
})
