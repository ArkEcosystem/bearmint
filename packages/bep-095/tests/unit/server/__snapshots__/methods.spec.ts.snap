// Vitest Snapshot v1

exports[`Server > should complete an M-of-N scheme 1`] = `
{
  "muSig": {
    "m": 2,
    "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
    "n": 3,
    "slaves": [
      "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
      "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
      "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
    ],
  },
  "signatures": [
    {
      "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
      "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
      "signature": "b12e84192c4b0a1aa298c120b7c077f95bb46393322f4e48514e33c870e32fa93f1316b8a03a38f87c5c3ad350646c2e0da08b6f31efbed3eb43d5e99d0454d017a5396b940ad7fdb47e6bcdc8f438965a234f6b3a9f8e80c354892b51d1e1d5",
    },
    {
      "id": "09bba3b4ea213f652d4718747622aeb2e17846c79ed8bc15a6ea18135f94e4e3",
      "publicKey": "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
      "signature": "a79a90b1432d9d4f51fee6f7a9be8133e88204beadc1fdec22bddacfac3ea4baf636120f20eed1711591723a02b1e10f16a379f99a05b1f80dd6b56d4a5231459614b239e078c6a0e8353e399c8e199145a4854a7ad8004dc87aef49079dec54",
    },
  ],
}
`;

exports[`Server > should complete an N-of-N scheme 1`] = `
{
  "muSig": {
    "m": 3,
    "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
    "n": 3,
    "slaves": [
      "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
      "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
      "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
    ],
  },
  "signatures": [
    {
      "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
      "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
      "signature": "b12e84192c4b0a1aa298c120b7c077f95bb46393322f4e48514e33c870e32fa93f1316b8a03a38f87c5c3ad350646c2e0da08b6f31efbed3eb43d5e99d0454d017a5396b940ad7fdb47e6bcdc8f438965a234f6b3a9f8e80c354892b51d1e1d5",
    },
    {
      "id": "09bba3b4ea213f652d4718747622aeb2e17846c79ed8bc15a6ea18135f94e4e3",
      "publicKey": "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
      "signature": "a79a90b1432d9d4f51fee6f7a9be8133e88204beadc1fdec22bddacfac3ea4baf636120f20eed1711591723a02b1e10f16a379f99a05b1f80dd6b56d4a5231459614b239e078c6a0e8353e399c8e199145a4854a7ad8004dc87aef49079dec54",
    },
    {
      "id": "267d857c3a01f739055a34028f1c2051769b0ead4ef33a0eeea06be81a8ba503",
      "publicKey": "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      "signature": "a0d64c09e49dc381fe59b24ee5bc80d21ee1fdbee9a1e49b501d98ef7fec1d909bf825e7e43f8ed616e2258404c78096174f7c173ad481c4cf7f1c4af76b4ae30b9828e6547523eb51b855af1d9ac6a7447e6e9e8b74a5d17966917113e63213",
    },
  ],
}
`;

exports[`Server > should list all txs 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "b12e84192c4b0a1aa298c120b7c077f95bb46393322f4e48514e33c870e32fa93f1316b8a03a38f87c5c3ad350646c2e0da08b6f31efbed3eb43d5e99d0454d017a5396b940ad7fdb47e6bcdc8f438965a234f6b3a9f8e80c354892b51d1e1d5",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Server > should list pending txs 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "b12e84192c4b0a1aa298c120b7c077f95bb46393322f4e48514e33c870e32fa93f1316b8a03a38f87c5c3ad350646c2e0da08b6f31efbed3eb43d5e99d0454d017a5396b940ad7fdb47e6bcdc8f438965a234f6b3a9f8e80c354892b51d1e1d5",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Server > should list ready txs 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "b12e84192c4b0a1aa298c120b7c077f95bb46393322f4e48514e33c870e32fa93f1316b8a03a38f87c5c3ad350646c2e0da08b6f31efbed3eb43d5e99d0454d017a5396b940ad7fdb47e6bcdc8f438965a234f6b3a9f8e80c354892b51d1e1d5",
      },
      {
        "id": "09bba3b4ea213f652d4718747622aeb2e17846c79ed8bc15a6ea18135f94e4e3",
        "publicKey": "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "signature": "a79a90b1432d9d4f51fee6f7a9be8133e88204beadc1fdec22bddacfac3ea4baf636120f20eed1711591723a02b1e10f16a379f99a05b1f80dd6b56d4a5231459614b239e078c6a0e8353e399c8e199145a4854a7ad8004dc87aef49079dec54",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;
